blocks/search.py

```diff
@@ -0,0 +1,266 @@
+from collections import OrderedDict
+from six.moves import range
+
+import numpy
+
+from theano import config
+from theano import function
+from theano import tensor
+
+from blocks.filter import VariableFilter
+from blocks.graph import ComputationGraph
+
+floatX = config.floatX
+
+
+def construct_dict(func):
+    def wrapper(self, *inputs):
+        names, outputs = func(self, *inputs)
+        return OrderedDict(zip(names, outputs))
+    return wrapper
+
+
+class BeamSearch(object):
+    """Beam search.
+
+    Parameters
+    ----------
+    beam_size : int
+        Size of beam.
+    sequence_generator : sequence generator
+        Sequence generator brick.
+    attended : theano variable
+        Theano variable for attended in sequence generator.
+    attended_mask : theano variable
+        Theano variable for attended mask in sequence generator.
+    inputs_dict : dict
+        Dictionary of inputs {name: theano variable for input}. The
+        functions will be constructed with these inputs.
+
+    """
+    def __init__(self, beam_size, sequence_generator, attended,
+                 attended_mask, inputs_dict):
+        self.beam_size = beam_size
+        self.sequence_generator = sequence_generator
+        self.attended = attended
+        self.attended_mask = attended_mask
+        self.inputs_dict = inputs_dict
+        self.generate_names = sequence_generator.generate.outputs
+        self.init_computer = None
+        self.next_computer = None
+        self.attended_computer = None
+        self.initial_state_computer = None
+        self.state_names = (sequence_generator.state_names +
+                            sequence_generator.glimpse_names +
+                            ['outputs'])
+        self.compiled = False
+
+    def compile(self, *args, **kwargs):
+        """Compiles functions for beam search."""
+        generator = self.sequence_generator
+        attended = self.attended
+        attended_mask = self.attended_mask
```

Comment:
Forgive if I am wrong, but it seems to me that instead of the pair `attended` and `attended_mask` you could as well work with just the list of contexts everywhere. That would kill two rabbits (of course there is no such an idiom in English..): the code will become both simpler and more general.